presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSource.java

```diff
@@ -0,0 +1,383 @@
+
+    private boolean finished;
+    private long readTimeNanos;
+    private long completedBytes;
+
+    public PinotBrokerPageSource(PinotConfig pinotConfig, ConnectorSession session, PinotQueryGenerator.GeneratedPql brokerPql, List<PinotColumnHandle> columnHandles, PinotClusterInfoFetcher clusterInfoFetcher, ObjectMapper objectMapper)
+    {
+        this.pinotConfig = requireNonNull(pinotConfig, "pinot config");
+        this.brokerPql = requireNonNull(brokerPql, "broker");
+        this.clusterInfoFetcher = requireNonNull(clusterInfoFetcher, "cluster info fetcher");
+        this.columnHandles = ImmutableList.copyOf(columnHandles);
+        this.session = requireNonNull(session, "session");
+        this.objectMapper = requireNonNull(objectMapper, "object mapper");
+    }
+
+    private static Double parseDouble(String value)
+    {
+        try {
+            return Double.valueOf(value);
+        }
+        catch (NumberFormatException ne) {
+            switch (value) {
+                case PINOT_INFINITY:
+                case PINOT_POSITIVE_INFINITY:
+                    return PRESTO_INFINITY;
+                case PINOT_NEGATIVE_INFINITY:
+                    return PRESTO_NEGATIVE_INFINITY;
+            }
+            throw new PinotException(PinotErrorCode.PINOT_DECODE_ERROR, Optional.empty(), "Cannot decode double value from pinot " + value, ne);
+        }
+    }
+
+    private void setValue(Type type, BlockBuilder blockBuilder, String value)
+    {
+        if (value == null) {
+            blockBuilder.appendNull();
+            return;
+        }
+        if (!(type instanceof FixedWidthType) && !(type instanceof VarcharType)) {
+            throw new PinotException(PinotErrorCode.PINOT_UNSUPPORTED_COLUMN_TYPE, Optional.empty(), "type '" + type + "' not supported");
+        }
+        if (type instanceof FixedWidthType) {
+            completedBytes += ((FixedWidthType) type).getFixedSize();
+            if (type instanceof BigintType) {
+                type.writeLong(blockBuilder, parseDouble(value).longValue());
+            }
+            else if (type instanceof IntegerType) {
+                blockBuilder.writeInt(parseDouble(value).intValue());
+            }
+            else if (type instanceof TinyintType) {
+                blockBuilder.writeByte(parseDouble(value).byteValue());
+            }
+            else if (type instanceof SmallintType) {
+                blockBuilder.writeShort(parseDouble(value).shortValue());
+            }
+            else if (type instanceof BooleanType) {
+                type.writeBoolean(blockBuilder, Boolean.valueOf(value));
+            }
+            else if (type instanceof DecimalType || type instanceof DoubleType) {
+                type.writeDouble(blockBuilder, parseDouble(value));
+            }
+            else if (type instanceof TimestampType) {
+                type.writeLong(blockBuilder, Long.valueOf(value));
+            }
+            else {
```

Comment:
I prefer having the else, otherwise I have to carefully reason about whether the types are exclusive or not (which they are in this case but not always).